File: py/src/comms/JsonEncoder.py
from typing import Type
import json

from .Encoder import Encoder
from .jsonRpc import Request, Response, RequestOrResponse

class JsonEncoder(Encoder):

    def name(self) -> str:
        return 'json'

    def decode(self, message: bytes, cls: Type[RequestOrResponse]) -> RequestOrResponse:
        instance = cls()
        dic = json.loads(message)
        instance.__dict__.update(dic)
        return instance

    def encode(self, obj: RequestOrResponse) -> bytes:
        return json.dumps(obj, cls=JSONEncoderExtension).encode('utf8')


class JSONEncoderExtension(json.JSONEncoder):
    """
    Extension of the builtin JSON encoder to handle objects
    """

    def default(self, obj):
        return obj.__dict__

File: js/src/comms/JsonEncoder.ts
import Encoder from './Encoder'
import JsonRpcRequest from './JsonRpcRequest'
import JsonRpcResponse from './JsonRpcResponse'

/**
 * A [JSON](http://json.org/) message encoder
 */
export default class JsonEncoder extends Encoder {

  name (): string {
    return 'json'
  }

  encode (instance: JsonRpcRequest | JsonRpcResponse): string {
    return JSON.stringify(instance)
  }

  decode<Type = JsonRpcRequest | JsonRpcResponse> (message: string, constructor: new () => Type): Type {
    const instance = new constructor()
    const data = JSON.parse(message)
    Object.assign(instance, data)
    return instance
  }
}
